assets.js

```diff
@@ -0,0 +1,143 @@
+    this.cache = new Cache(options.cache);
+    this.skipCache = !options.cache.enabled;
+  }
+
+    /**
+   * Parses the parameters for get calls, normalizing into known names and applying the proper defaults & encodings
+   *
+   * @param {Object} params Parameters that specify pkg, env, version and/or locale.
+   * @returns {Object} The params parsed into `{ env, version, locale, pkg, filter }` with default values applied
+   * and `pkg` uri encoded
+   * @private
+   */
+  _readParams(params) {
+    const env = environments.get(params.environment || params.env) || 'dev';
+    const version = params.v || params.version;
+    const locale = params.locale || 'en-US';
+
+    let pkg = params.package || params.pkg;
+    if (pkg) {
+      //
+      // Handle scoped packages
+      //
+      pkg = encodeURIComponent(pkg);
+    }
+
+    let filter = params.filter;
+    if (filter) {
+      filter = encodeURIComponent(filter);
+    }
+    return { env, version, locale, pkg, filter };
+  }
+
+  /**
+   * Get asset information. Build information may be read from the cache if configured.
+   *
+   * @param {Object} params Parameters that specify pkg, env, version and/or locale.
+   * @param {Function} fn Completion callback.
+   * @returns {Warehouse} fluent interface.
+   * @public
+   */
+  get(params, fn) {
+    const { env, version, locale, pkg, filter } = this._readParams(params);
+
+    if (!pkg) {
+      return fn(new Error('invalid parameters supplied, missing `pkg`'));
+    }
+
+    const cacheKey = {
+      type: 'assets',
+      env,
+      version,
+      locale,
+      pkg,
+      filter
+    };
+
+    this.cache.get(JSON.stringify(cacheKey), { skipCache: this.skipCache || params.bypassCache }, key => {
+      return new Promise((resolve, reject) => {
+        this._get(JSON.parse(key), (err, data) => {
+          if (err) {
+            reject(err);
+            return;
+          }
+
+          debug('Caching asset data: pkg = %s, env = %s, version = %s', pkg, env, version);
+          resolve(data);
+        });
+      });
+    })
+    .then(data => {
+      debug('Returning cached asset data: pkg = %s, env = %s, version = %s', pkg, env, version);
+      fn(null, data.value);
+    })
+    .catch(fn);
+
+    return this.warehouse;
+  }
+
+  /**
+   * Get asset information. The cache is never used. This method assumes all the parameters have already been validated.
+   *
+   * @param {Object} options Options object describing the build to get
+   * @param {string} options.pkg The package name
+   * @param {string} options.env The environment
+   * @param {string} options.version The version
+   * @param {string} options.locale The locale
+   * @param {Function} fn Completion callback.
+   * @returns {Warehouse} fluent interface.
+   * @private
+   */
+  _get({ pkg, env, version, locale, filter }, fn) {
+    debug('Build metadata: pkg = %s, env = %s, version = %s', pkg, env, version, filter);
+    return this.warehouse.send(
```

Comment:
Maybe we should add a `Warehouse.prototype.sendify` (I'm not married to the name)

```javascript
Warehouse.prototype.sendify = function sendify(pathname, options) {
  return new Promise((resolve, reject) => {
    this.send(pathname, options, function(err, body) {
      if(err) return reject(err);

      resolve(body);
    });
  });
};
```

It would let us async await all this, and clean it up a bunch.

Reply:
yeah. I think we should consider just removing node@6 support and just come through and async..await everything.  But not now.  Maybe a ticket for the future.

Reply:
Apparently we already don't support [`node@6`](https://github.com/warehouseai/warehouse.ai-api-client/blob/master/package.json#L25)

I'd still rather tackle something like that more holistically and consider changing the api to return promises so callers also can do async..await